src/core/lib/channel/channelz.h

```diff
@@ -70,6 +70,23 @@ class CallCountingHelperPeer;
 class ChannelNodePeer;
 }  // namespace testing
 
+inline static const char* GetChannelConnectivityStateChangeString(
```

Comment:
Is there a reason this needs to be `inline`?  It doesn't seem like it's that performance sensitive.  It would be cleaner to put the definition in the .cc file, just to avoid making people read more code in the .h file that isn't really part of the API definition.

If we do keep this inline, then I don't think it also needs to be `static`.  (Does that even have any meaning with `inline`?)

Actually, better yet: Instead of making this a stand-alone function, how about making it a static member of `ChannelNode`, since that's the only place where it's used?

Reply:
Would it be better to make it a member function of ChannelNode rather than a static member? We are only ever using this after we have set the connectivity state on the ChannelNode, so it seems like ChannelNode can just give us the string version of its state whenever we need it.

I initially tried putting the definition in the .cc file, but got compilation errors due to linkage. I agree it's better not to have the definition in the .cc file. I don't know about static plus inline.

Reply:
I don't think it really makes sense to have it read the current state out of the `ChannelNode` object, since we're only going to be using right where we're setting the state.  I think it would make sense to do that only if we were going to have it automatically add the trace message whenever the state was set (in which case it would not have to be exposed outside of `ChannelNode` at all), but I think it should be up to the channel code as to whether or not to add the trace message.

Making it a static member seems better, because it does not actually need to access any state from the `ChannelNode` object.

> I initially tried putting the definition in the .cc file, but got compilation errors due to linkage. I agree it's better not to have the definition in the .cc file. I don't know about static plus inline.

I assume you meant ".h" instead of ".cc" there.

If you put the definition in the .h file without declaring it inline, then yeah, that will cause link errors.  You can make that go away by either making it inline or declaring it in the .h file but moving the definition to the .cc file.  I think the latter is cleaner unless there's a performance-related reason why it has to be inlined.

Reply:
Oops, yeah, I meant to say " I agree it's better not to have the definition in the *.h* file." I got linkage errors when putting the definition in the *.cc* file but I wasn't sure why.

Reply:
Was it defined `static` when you had it in the .cc file?  If so, that wouldn't work, because the symbol wouldn't have been exported.

What was the exact linker error you were seeing?

Reply:
That's probably what happened :) I don't remember the exact linker error because it was a few days ago, but I made your suggested change and it seems to work.